src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "security/gssapi_principal_mapper.h"
+
+#include "security/logger.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <boost/algorithm/string/case_conv.hpp>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <charconv>
+
+namespace security {
+static constexpr const char* const non_simple_pattern = R"([/@])";
+static constexpr const char* const parameter_pattern = R"(([^$]*)(\$(\d*))?)";
+
+gssapi_rule::gssapi_rule(std::string_view default_realm)
+  : _default_realm(default_realm) {}
+
+gssapi_rule::gssapi_rule(
+  std::string_view default_realm,
+  int number_of_components,
+  std::string_view format,
+  std::string_view match,
+  std::string_view from_pattern,
+  std::string_view to_pattern,
+  repeat repeat_,
+  make_lower to_lower,
+  make_upper to_upper)
+  : _default_realm(default_realm)
+  , _is_default(false)
+  , _number_of_components(number_of_components)
+  , _format(format)
+  , _match(match)
+  , _from_pattern(std::regex{
+      from_pattern.begin(),
+      from_pattern.length(),
+      std::regex_constants::ECMAScript | std::regex_constants::optimize})
+  , _from_pattern_str(from_pattern)
+  , _to_pattern(to_pattern)
+  , _repeat(repeat_)
+  , _to_lower(to_lower)
+  , _to_upper(to_upper) {}
+
+std::optional<ss::sstring>
+gssapi_rule::apply(std::vector<std::string_view> params) const {
+    static const RE2 non_simple_regex(non_simple_pattern, RE2::Quiet);
+    const re2::StringPiece match_piece(_match.data(), _match.size());
+    const RE2 match_regex(match_piece, RE2::Quiet);
```

Comment:
should we use the scoped `re2::RE2`? it's weird that they export RE2 into the global scope in the headers...

Reply:
I can do that for sure.  I agree it is weird, and not a fan of 3rd party libraries bringing things into scope outside of my control.